<!doctype HTML PUBLIC "-//W3C//DTD HTML 4.01//EN" "http://www.w3.org/TR/html4/strict.dtd">
<html lang="EN">
<!-- ABAPCATCH_TRY -->

<!-- Mirrored from help.sap.com/doc/abapdocu_754_index_htm/7.54/en-US/abapcatch_try.htm by HTTrack Website Copier/3.x [XR&CO'2014], Sun, 18 Feb 2024 14:51:36 GMT -->
<head>
<meta http-equiv="content-type" content="text/html; charset=utf-8">
<meta name="author" content="SAP">
<meta name="description" content="CATCH">
<meta name="keywords" content="ABAP, CATCH">
<title>
CATCH - ABAP Keyword Documentation
</title>
<link rel="icon" href="ABAPIcon.ico" type="image/ico">
<link rel="shortcut icon" href="ABAPIcon.ico">
<LINK rel="stylesheet" href="abap_docu.css" type="text/css">
<script language="javascript" type="text/javascript" src="functions.js"></script>
<script type="text/JavaScript">
function urlToClipboard(id){
 try{ var loc = top.location.href;}
 catch(e){ alert( "URL cannnot be determined" ); return; }
 var field = document.getElementById(id);
 field.focus();
 if ( loc.search( ".htm" ) == -1 && loc.search( ".HTM" ) == -1  ){ loc = loc + "index.htm"; }
 if (   loc.search( "index.html" ) != -1
     || loc.search( "INDEX-2.html" ) != -1 ){
   var off = loc.search( "file=" );
   if ( off == -1  ){
     field.value = loc + "?file=abapcatch_try.htm"; }
   else {
     field.value = loc.substring( 0, off ) + "file=abapcatch_try.htm";
   }
 }
 else {
    field.value = loc; }
 field.setSelectionRange(0, field.value.length);
 var r = confirm( "Copy URL to clipboard?\n\n" + field.value + "\n\nCopy to clipboard does not work in all browsers." );
 if (r == true) {
   try{
     document.execCommand("copy");
   }
     catch(e){ alert( "URL could not be copied to clipboard" ); }
 }
 window.scrollTo(0,0);
}
</script>
</head>
<body>
<div class="all">
<div style="float:right;">
<input type="button" value="URL"
style="font-size:10px; width:30px; height:18px; padding:0; "
onclick="urlToClipboard('url')" title="Copy URL to Clipboard" ></div>
<p>
<input name="query" id="query" value="" type="text" title="Suchbegriff" size="30" maxlength="30" onkeyup="getInput()">&nbsp;
<input type="button" value="Search"
style="font-size:12px; width:50px; height:18px; padding:0"
onclick="call_search('search.html')" title="Search in ABAP Keyword Documentation" >
<hr style="width:100%;height:1px"></p>
<p class="copyright">AS ABAP Release 754, &copy;Copyright 2019 SAP SE. All rights reserved.</p>
<span class="path">
<a href="javascript:call_link('abenabap.html')" class="blue">ABAP Keyword Documentation</a>&nbsp;&rarr;&nbsp;
<a href="javascript:call_link('abenabap_reference.html')" class="blue">ABAP &#x2212; Reference</a>&nbsp;&rarr;&nbsp;
<a href="javascript:call_link('abenabap_flow_logic.html')" class="blue">Program Flow Logic</a>&nbsp;&rarr;&nbsp;
<a href="javascript:call_link('abenabap_exceptions.html')" class="blue">Exception Handling</a>&nbsp;&rarr;&nbsp;
<a href="javascript:call_link('abenexceptions.html')" class="blue">Class-Based Exceptions</a>&nbsp;&rarr;&nbsp;
<a href="javascript:call_link('abaptry.html')" class="blue">TRY</a>&nbsp;&rarr;&nbsp;
</span>
<A NAME="@@CATCH@@!ABAP_STATEMENT!@@"></A>
<p><span class="h1">
CATCH <br><br>
</span></p>
<P>
<a href="javascript:call_link('abaptry_shortref.html')" class="blue">Quick Reference</a> <br><br>
</P>
<p><span class="h2">
<span class="bold">Syntax</span>
</span></p>
<P>
<span class="qtext">CATCH <i class=bn>[</i>BEFORE UNWIND<i class=bn>]</i> cx_class1 cx_class2 ... <i class=bn>[</i>INTO oref<i class=bn>]</i>.</span> <br><br>
</P>
<p><span class="h3">
Extras:
</span></p>
<P>
   <A HREF="#!ABAP_ADDITION_1@1@" class="blue"> 1. <span class="qtext">... BEFORE UNWIND</span> </A> <br><br><A HREF="#!ABAP_ADDITION_2@2@" class="blue"> 2. <span class="qtext">... INTO oref</span> </A> <br><br>
</P>
<p><span class="h4">
Effect
</span></p>
<P>
Introduces a <span class="qtext">CATCH</span> block of a <a href="javascript:call_link('abaptry.html')" class="blue"><span class="qtext">TRY</span></a> control structure in which exceptions can be handled.
</P>
<P>
 A <span class="qtext">CATCH</span> block is an exception handler, meaning the program logic that is
executed whenever the associated exception is raised in the <span class="qtext">TRY</span> block of the same <span class="qtext">TRY</span> control structure.
</P>
<P>
A <span class="qtext">CATCH</span> block handles the exceptions of the exception classes <span class="qtext">cx_class1
cx_class2 ...</span> that are specified after the statement <span class="qtext">CATCH</span> as well
as the exceptions of the subclasses of these exception classes. In each <span class="qtext">CATCH</span>
statement of a <span class="qtext">TRY</span> control structure, any number of exception classes <span class="qtext">
cx_class1 cx_class2 ...</span> can be specified, with more specific exception classes (subclasses) being
specified before more general exception classes (superclasses). This order must be preserved both within
a <span class="qtext">CATCH</span> statement and across multiple <span class="qtext">CATCH</span> statements of a <span class="qtext">TRY</span> control structure.
</P>
<P>
Before the <span class="qtext">CATCH</span> block is executed, the system by default deletes the context
in which the exception was raised. To preserve the context during execution of the <span class="qtext">CATCH</span> block, the <span class="qtext">BEFORE UNWIND</span> addition can be used. <br><br>
</P>
<p><span class="h4">
Note
</span></p>
<P>
The rule where special exception classes must be specified before general classes in <span class="qtext">CATCH</span>
ensures that an exception is not handled by a general exception handler (superclass) if a special handler (subclass) is provided. <br><br>
</P>
<p><span class="h4">
Example
</span></p>
<p>
Catch two possible exceptions with <span class="qtext">CATCH</span>. If the input cannot be interpreted
as number, the exception CX_SY_CONVERSION_NO_NUMBER is caught by its superclass CX_SY_CONVERSION_ERROR. If the number 0 is entered, the exception CX_SY_CONVERSION_ERROR is caught by its superclass CX_SY_ARITHMETIC_ERROR.
</p>
<p class="qtextml1">
<span class="qtext">
DATA(input) = `x`. <br>cl_demo_input=>request( CHANGING field = input ). <br> <br>TRY. <br>&nbsp;&nbsp;&nbsp;&nbsp;cl_demo_output=>display( |{ 1 / CONV decfloat34( input ) }| ). <br>
&nbsp;&nbsp;CATCH cx_sy_arithmetic_error cx_sy_conversion_error INTO DATA(exc). <br>&nbsp;&nbsp;&nbsp;&nbsp;cl_demo_output=>display( exc->get_text( ) ). <br>ENDTRY. <br><br>
</span>
</p>
<p><span class="h4">
Executable Example
</span></p>
<P>
<a href="javascript:call_link('abencatch_exception_abexa.html')" class="blue">Exceptions, <span class="qtext">CATCH</span></a> <br><br>
</P>
<A NAME="@@BEFORE UNWIND@@CATCH@@"></A>
<p><span class="h3">
<A NAME="!ABAP_ADDITION_1@1@"></A> <span class="bold">Addition 1</span>
</span></p>
<P>
<span class="qtext">... BEFORE UNWIND</span>   <br><!-- /A -->
</P>
<p><span class="h4">
Effect
</span></p>
<P>
If the addition <span class="qtext">BEFORE UNWIND</span> is specified, the context in which the exception
was raised is not deleted until the <span class="qtext">CATCH</span> block is deleted. Instead, the
context is preserved (including all called procedures and their local data) during the execution of the <span class="qtext">CATCH</span> block.
</P>
<ul class="disc">
<li>
If no <a href="javascript:call_link('abapresume.html')" class="blue"><span class="qtext">RESUME</span></a> statement is executed
in the <span class="qtext">CATCH</span> block, the context is deleted when the <span class="qtext">CATCH</span> block is exited.
<br><br></li>
<li>
Of a <a href="javascript:call_link('abapresume.html')" class="blue"><span class="qtext">RESUME</span></a> statement is executed in the <span class="qtext">CATCH</span> block, processing resumes after the place that raised the exception.
<br></li>
</ul>
<P>
Any <a href="javascript:call_link('abapcleanup.html')" class="blue"><span class="qtext">CLEANUP</span></a> blocks are always
executed directly before their context is deleted. If <span class="qtext">BEFORE UNWIND</span> is used,
after exception handling, and in all other cases before the exception handling. In a <span class="qtext">CATCH</span>
block with <span class="qtext">BEFORE UNWIND</span>, no statements can be executed in which the context is closed without executing any <span class="qtext">CLEANUP</span> blocks.
</P>
<ul class="disc">
<li>
The statements <a href="javascript:call_link('abapcall_transaction.html')" class="blue"><span class="qtext">CALL TRANSACTION</span></a>,
<a href="javascript:call_link('abapcall_dialog.html')" class="blue"><span class="qtext">CALL DIALOG</span></a>, <a href="javascript:call_link('abapcall_transaction.html')" class="blue">
 <span class="qtext">SUBMIT</span></a>, <a href="javascript:call_link('abapstop.html')" class="blue"><span class="qtext">STOP</span></a>,
<a href="javascript:call_link('abapreject.html')" class="blue"><span class="qtext">REJECT</span></a>, <a href="javascript:call_link('abapcheck_select-options.html')" class="blue">
<span class="qtext">CHECK SELECT-OPTIONS</span></a>, and the obsolete variant of <a href="javascript:call_link('abapleave-.html')" class="blue"><span class="qtext">LEAVE</span></a> are forbidden, as they might leave the
<a href="javascript:call_link('abeninternal_session_glosry.html')" class="grey" title="Glossary Entry">internal session</a>.
<br><br></li>
<li>
Statements such as <a href="javascript:call_link('abapleave_to_transaction.html')" class="blue"><span class="qtext">LEAVE TO
TRANSACTION</span></a>, which are statically known to leave the internal session, are permitted. In
such cases, exception handling is correctly ended while executing the <span class="qtext">CLEANUP</span> blocks before the statement is executed.
<br><br></li>
<li>
Procedure calls are allowed. However, if the context is deleted during this type of procedure call,
a runtime error occurs EXCP_HANDLER_FAILED_TO_UNWIND. This is also the case if a statement is executed
here, which is allowed in the <span class="qtext">CATCH</span> block. For example, <a href="javascript:call_link('abapleave_to_transaction.html')" class="blue">
 <span class="qtext">LEAVE TO TRANSACTION</span></a> can be executed directly in the <span class="qtext">CATCH</span>
block, but not in a procedure that is called in this block. This is because the context would be deleted if the <span class="qtext">CLEANUP</span> blocks are not executed.
<br><br></li>
<li>
If the <a href="javascript:call_link('abenmessage_type_glosry.html')" class="grey" title="Glossary Entry">message type</a> is
dynamically specified for the <a href="javascript:call_link('abapmessage.html')" class="blue"><span class="qtext">MESSAGE</span></a>
statement for sending messages, the ABAP runtime environment behaves as if exception handling is going
to be exited. The context is deleted during the execution of the <span class="qtext">CLEANUP</span>
blocks. However, if the system returns to the <span class="qtext">CATCH</span> block after the <span class="qtext">
MESSAGE</span> statement has been executed (which can be the case with message types I and S, for example), the exception CX_SY_ILLEGAL_HANDLER is raised.
<br></li>
</ul>
<P>
When combined with <span class="qtext">INTO</span>, the addition <span class="qtext">BEFORE UNWIND</span>
sets the attribute IS_RESUMABLE of the exception object and of the preceding exception objects in a
chaining with the attribute PREVIOUS. Up until the first resumable raised exception, IS_RESUMABLE is set to the value of ABAP_TRUE and is otherwise set to the value of ABAP_FALSE. <br><br>
</P>
<p><span class="h4">
Notes
</span></p>
<ul class="disc">
<li>
If addition <span class="qtext">BEFORE UNWIND</span>is not specified, the context is deleted before the <span class="qtext">CATCH</span> block is executed.
<br><br></li>
<li>
The statement <a href="javascript:call_link('abapresume.html')" class="blue"><span class="qtext">RESUME</span></a> can be used only when handling a
<a href="javascript:call_link('abenresumable_exception_glosry.html')" class="grey" title="Glossary Entry">resumable exception</a>
and only in a <span class="qtext">CATCH</span> block in which the addition <span class="qtext">BEFORE
UNWIND</span> is specified. This is the only case in which the context of the exception is not deleted when the <span class="qtext">CATCH</span> block is exited. <br>
<br></li>
<li>
Resumable exceptions can also be handled in <span class="qtext">CATCH</span> blocks without the addition
<span class="qtext">BEFORE UNWIND</span>. In this case, the context of the exception is deleted before the handling process and the statement <span class="qtext">RESUME</span> cannot be specified. <br>
<br></li>
<li>
Use of the addition <span class="qtext">BEFORE UNWIND</span> for <span class="qtext">CATCH</span> is
only required when the statement <a href="javascript:call_link('abapresume.html')" class="blue"><span class="qtext">RESUME</span></a>
is used. However, it is allowed in principle during exception handling if the context of the exception
is to be evaluated before any cleanup activities in <a href="javascript:call_link('abapcleanup.html')" class="blue"><span class="qtext">CLEANUP</span></a>
blocks. This makes sense, for example, when handling resource bottlenecks if releasing resources in
<span class="qtext">CLEANUP</span> blocks would change the context and thus make the calculation of
the free resources in the exception handler meaningless. Other than for logging purposes, we do not
recommend evaluating the part of the context that is only of interest locally for implementing the incorrect procedure. <br>
<br></li>
<li>
In a procedure called up using <span class="qtext">BEFORE UNWIND</span> during a <span class="qtext">CATCH</span>
block, each <span class="qtext">MESSAGE</span> statement that sends a message causes the runtime error EXCP_HANDLER_FAILED_TO_UNWIND. <br><br>
<br></li>
</ul>
<p><span class="h4">
Example
</span></p>
<p>
Uses the addition <span class="qtext">BEFORE UNWIND</span> when a resumable exception <span class="qtext">cx_demo</span> is caught.
</p>
<p class="qtextml1">
<span class="qtext">
CLASS cx_demo DEFINITION INHERITING FROM cx_static_check. <br>ENDCLASS. <br> <br>CLASS cls DEFINITION. <br>
&nbsp;&nbsp;PUBLIC SECTION. <br>&nbsp;&nbsp;&nbsp;&nbsp;CLASS-METHODS meth RAISING RESUMABLE(cx_demo). <br>
ENDCLASS. <br> <br>CLASS cls IMPLEMENTATION. <br>&nbsp;&nbsp;METHOD meth. <br>&nbsp;&nbsp;&nbsp;&nbsp;... <br>
&nbsp;&nbsp;&nbsp;&nbsp;RAISE RESUMABLE EXCEPTION TYPE cx_demo. <br>&nbsp;&nbsp;&nbsp;&nbsp;cl_demo_output=>display( 'Resumed ...' ). <br>
&nbsp;&nbsp;&nbsp;&nbsp;... <br>&nbsp;&nbsp;ENDMETHOD. <br>ENDCLASS. <br> <br>START-OF-SELECTION. <br>&nbsp;&nbsp;TRY. <br>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;cls=>meth( ). <br>&nbsp;&nbsp;&nbsp;&nbsp;CATCH BEFORE UNWIND cx_demo. <br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;RESUME. <br>&nbsp;&nbsp;ENDTRY. <br><br>
</span>
</p>
<A NAME="@@INTO@@CATCH@@"></A>
<p><span class="h3">
<A NAME="!ABAP_ADDITION_2@2@"></A> <span class="bold">Addition 2</span>
</span></p>
<P>
<span class="qtext">... INTO oref</span>   <br><!-- /A -->
</P>
<p><span class="h4">
Effect
</span></p>
<P>
If the addition <span class="qtext">INTO</span> is specified, a reference to the exception object is saved to <span class="qtext">oref</span>. The following can be specified for <span class="qtext">oref</span>:
</P>
<ul class="disc">
<li>
An existing object reference variable <span class="qtext">oref</span>, whose static type must be more general or as general as the most general of the specified exception classes.
<br><br></li>
<li>
An inline declaration with <a href="javascript:call_link('abendata_inline.html')" class="blue"><span class="qtext">DATA(var)</span></a>.
The static type of the declared object reference variable is the exception class (if specified). If
multiple exception classes are specified and a common superclass of these classes is used, the superclass is the static type of <span class="qtext">oref</span>; if not, it is CX_ROOT.
<br></li>
</ul>
<P>
The object reference variable can be used to access the attributes and methods of the exception object.
</P>
<P>
If the exception was raised with <a href="javascript:call_link('abapraise_exception_class.html')" class="blue"><span class="qtext">RAISE
RESUMABLE EXCEPTION</span></a> and the addition <span class="qtext">BEFORE UNWIND</span> was specified,
the attribute IS_RESUMABLE of the current exception object and of any preceding exception objects referenced
in the attribute PREVIOUS is, up until the first resumable raised exception, set to the value of ABAP_TRUE. In all other exception objects, the attribute IS_RESUMABLE is set to the value of ABAP_FALSE.
</P>
<p><span class="h4">
Note
</span></p>
<P>
Within a <span class="qtext">CATCH</span> block without the addition <span class="qtext">BEFORE UNWIND</span>, the attribute IS_RESUMABLE of the exception object is undefined.
</P>
<p><span class="h4">
Example
</span></p>
<P>
Catches exceptions with an inline declaration of an object reference variable. The static type of this variable is <span class="qtext">cx</span>.
</P>
<p class="qtextml1">
<span class="qtext">
CLASS cx DEFINITION INHERITING FROM cx_dynamic_check. <br>&nbsp;&nbsp;PUBLIC SECTION. <br>ENDCLASS. <br>
 <br>CLASS cy DEFINITION INHERITING FROM cx. <br>&nbsp;&nbsp;PUBLIC SECTION. <br>ENDCLASS. <br> <br>TRY. <br>&nbsp;&nbsp;&nbsp;&nbsp;... <br>&nbsp;&nbsp;CATCH cy cx INTO DATA(oref). <br>ENDTRY.
</span>
</p>
<br>
<br>
<br>
<br>
</DIV>
<fieldset style="opacity:0">
<input id="url" type="text" />
</fieldset>
</BODY>

<!-- Mirrored from help.sap.com/doc/abapdocu_754_index_htm/7.54/en-US/abapcatch_try.htm by HTTrack Website Copier/3.x [XR&CO'2014], Sun, 18 Feb 2024 14:51:36 GMT -->
</html>
